react-components/src/analysis/view/AnalysesView.js

```diff
@@ -584,35 +599,59 @@ class AnalysesView extends Component {
         }
     }
 
-    handleRowClick(id) {
+    handleRowClick(index) {
         this.setState((prevState, props) => {
-            const { selected } = prevState;
+            const { data, selected } = prevState;
+            const id = data[index].id;
+
+            const newState = { lastSelectedIndex: index };
+
             if (selected.indexOf(id) < 0) {
-                return { selected: [id] };
+                newState.selected = [id];
             }
+
+            return newState;
         });
     }
 
-    handleCheckBoxClick(event, id) {
+    handleCheckBoxClick(event, index) {
+        const selectRange = event.shiftKey;
+
         this.setState((prevState, props) => {
-            const { selected } = prevState;
-            const selectedIndex = selected.indexOf(id);
+            const { data, selected, lastSelectedIndex } = prevState;
+            const id = data[index].id;
+            const isSelected = this.isSelected(id);
+
             let newSelected = [];
 
-            if (selectedIndex === -1) {
-                newSelected = newSelected.concat(selected, id);
-            } else if (selectedIndex === 0) {
-                newSelected = newSelected.concat(selected.slice(1));
-            } else if (selectedIndex === selected.length - 1) {
-                newSelected = newSelected.concat(selected.slice(0, -1));
-            } else if (selectedIndex > 0) {
-                newSelected = newSelected.concat(
-                    selected.slice(0, selectedIndex),
-                    selected.slice(selectedIndex + 1)
-                );
+            if (selectRange && lastSelectedIndex !== index) {
+                const allIDs = data.map((analysis) => analysis.id);
+
+                newSelected =
+                    lastSelectedIndex < index
+                        ? rangeSelect(
+                              allIDs,
+                              selected,
+                              isSelected,
```

Comment:
I'm adding or removing everything in the range solely based on whether the last item in the range was already selected or not. 

In other words, if you shift+click on an item that wasn't already selected, then everything in the range becomes selected, even if some of them already were selected. If the last item was already selected, then everything in the range becomes deselected, even if only one other item in the range was selected.

This is also how Gmail behaves.